core/arch/arm/kernel/boot.c

```diff
@@ -1266,6 +1273,29 @@ void __weak boot_init_primary_early(void)
 		mapped_tl = transfer_list_map(boot_arg_transfer_list);
 		if (!mapped_tl)
 			panic("Failed to map transfer list");
+
+		transfer_list_dump(mapped_tl);
+		tl_e = transfer_list_find(mapped_tl, TL_TAG_FDT);
```

Comment:
Yes. If it is the last entry, we don't need to worry about its next TE being corrupted by adding new DTB nodes later.

Reply:
Fixed.

Reply:
If the TE is the last one, we can keep `data_size` unchanged here, it reflects the actual size of the TE data. When OP-TEE inserting new nodes, it can just expand its size at that time and no worries on overriding other TEs followed by.
The advantage is we can skip the `memmove` in `transfer_list_set_data_size()` when it is the last TE.

Reply:
That optimization seems a bit dangerous. What if another entry is added to the transfer list after this? I assume it doesn't happen right now, but since the functions are there it might only be a matter of time before it's attempted.

Reply:
@jenswi-linaro 
We can reserve the space for DTB TE in both cases (without checking `transfer_list_next(mapped_tl, tl_e)` as @etienne-lms suggested).
However, actually adding a new TE is still not allowed after `boot_init_primary_early()` and before calling `release_transfer_list()` because DTB TE takes up all the available space. Calling `transfer_list_add()` just return NULL.

A solution can be not allowing appending new nodes into the DTB but only putting new nodes as overlays in the form of a new TE. In this case we don't need to reserve the space for DTB TE during early boot.
But this needs to reflect changes into the FW Handoff spec and other components like U-Boot and hardly to apply in this patch.

Reply:
> However, actually adding a new TE is still not allowed after boot_init_primary_early() and before calling release_transfer_list() because DTB TE takes up all the available space. Calling transfer_list_add() just return NULL.

Yes, and that's fine, much better than accepting it and silently corrupting the DTB.

Reply:
Done.